metals/src/main/scala/scala/meta/internal/metals/debug/server/MainClassDebugAdapter.scala

```diff
@@ -0,0 +1,204 @@
+      if (msg.startsWith(JDINotificationPrefix)) {
+        if (initialized.compareAndSet(false, true)) {
+          val port = Integer.parseInt(msg.drop(JDINotificationPrefix.length))
+          val address = new InetSocketAddress("127.0.0.1", port)
+          listener.onListening(address)
+        }
+      } else {
+        listener.out(msg)
+      }
+    }
+
+    // Note that we current only shorten the classpath portion and not other options
+    // Thus we do not yet *guarantee* that the command will not exceed OS limits
+    val process =
+      if (cmdLength <= processCmdCharLimit) {
+        SystemProcess.run(
+          cmd,
+          root,
+          redirectErrorOutput = false,
+          envOptions,
+          processErr = Some(logError),
+          processOut = Some(logOutput),
+        )
+      } else {
+        Utils.withTempManifestJar(classPath) { manifestJar =>
+          val shortClasspathOption = "-cp" :: manifestJar.syntax :: Nil
+          val shortCmd =
+            java :: jvmOptions ::: shortClasspathOption ::: appOptions
+          SystemProcess.run(
+            shortCmd,
+            root,
+            redirectErrorOutput = false,
+            envOptions,
+            processErr = Some(logError),
+            processOut = Some(logOutput),
+          )
+        }
+      }
+
+    new CancelableFuture[Unit] {
+      def future = process.complete.map { code =>
+        if (code != 0)
+          throw new Exception(s"debuggee failed with error code $code")
+      }
+      def cancel(): Unit = process.cancel
+    }
+  }
+
+  private def enableDebugInterface: String = {
+    s"-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,quiet=n"
+  }
+}
+
+object Utils {
+  def withTempManifestJar(
+      classpath: Seq[Path]
+  )(
+      op: AbsolutePath => SystemProcess
+  )(implicit ec: ExecutionContext): SystemProcess = {
+
+    val manifestJar =
+      Files.createTempFile("jvm-forker-manifest", ".jar").toAbsolutePath
+    val manifestJarAbs = AbsolutePath(manifestJar)
+
+    // Add trailing slash to directories so that manifest dir entries work
+    val classpathStr =
+      classpath.map(addTrailingSlashToDirectories).mkString(" ")
+
+    val manifest = new Manifest()
+    manifest.getMainAttributes.put(Attributes.Name.MANIFEST_VERSION, "1.0")
+    manifest.getMainAttributes.put(Attributes.Name.CLASS_PATH, classpathStr)
+
+    val zipOut = new ZipOutputStream(Files.newOutputStream(manifestJar))
+    try {
+      val zipEntry = new ZipEntry(JarFile.MANIFEST_NAME)
+      zipOut.putNextEntry(zipEntry)
+      manifest.write(new BufferedOutputStream(zipOut))
+      zipOut.closeEntry()
+    } finally {
+      zipOut.close()
+    }
+
+    val process = op(manifestJarAbs)
+    process.complete.onComplete { case _ =>
+      manifestJarAbs.delete()
+    }
+    process
+  }
+
+  private def addTrailingSlashToDirectories(path: Path): String = {
+    // NOTE(olafur): manifest jars must use URL-encoded paths.
```

Comment:
Is this note from Bloop? Let's avoid mentioning olafur since the comment was added some where else (super nitpick I know)